### ISSUE TITLE
PPO ratio

### ISSUE THREAD

**Issue by TomasMerva** [openai/spinningup#181]
I am trying to implement PPO but I have come to few problems.
1. What exactly is the difference between logp, logp_pi? I understand it like this: the logp is probability of action taken in a state. The logp_pi is output from the same neural network with same states (collected from agent's interaction in an environment) but the selected action can differ then action, which was actually selected in logp [ the probabilities (outputs from neural network) are same but the action is randomly( based on those probabilities) selected ]. Do I understand it correctly?
2. The second question is why do we use logp_pi anyways? According to the Spinningup tutorial and also the PPO publication, we should use as ratio the probabilities of taking action in same states with old and new policy. (I suppose that states have to be same, but the action can be different, am I right?). In this case, should I make 2 neural networks, one representing new policy and the other one old policy? Therefore the algorithm should be like this:
1. Collect trajectories, compute advantage function,...
2. Make an update of the second (minor) neural network using Vanilla Policy gradient [ log_pi(at|st) * At ]
3. Feed the updated second neural network with same states (collected before)
4. Compute PPO update (the aforementioned ratio and clipped thing) 
5. The second network == the first network
6. Use PPO update for updating the first (major) neural network
7. update Value function approximator and repeat...
Is this right solution?

`ratio = tf.exp(logp - logp_old_ph)          # pi(a|s) / pi_old(a|s)
    min_adv = tf.where(adv_ph>0, (1+clip_ratio)*adv_ph, (1-clip_ratio)*adv_ph)
    pi_loss = -tf.reduce_mean(tf.minimum(ratio * adv_ph, min_adv))
    v_loss = tf.reduce_mean((ret_ph - v)**2)`

Using this version of ratio seems to not be based on the pseudocode I suppose. I just dont understand why do we use this version, or more specifically why logp_pi does represent old_policy. Thank you all for your help in advance.

**Comment by hanjialeOK** [comment]
I am reading spinningup/trpo and I have the exact same question with you at first. But now I Know it. Actually, `pi` is exactly the action taken in a state, so `logp_pi` is the right probability from old params.
